MAINT: backports for SciPy 1.16.2

tylerjereddy opened the pull request:
We're about a week away from targeted `1.16.2` release date--getting this rolling to see what kinds of problems arise early on.

Backports included so far:

1. gh-23122
2. gh-23133
3. gh-23403
4. gh-23414
5. gh-23435
6. gh-23442
7. gh-23452
8. gh-23457
9. gh-23471
10. gh-23475
11. gh-23488
12. gh-23495
13. gh-23540
14. gh-23154 (OpenBLAS bump only)
15. gh-23576
16. gh-23563


TODO:

- [x] update release notes, including mention of WoA binaries
- [x] regular CI passing
- [x] wheels CI passing, including new WoA binary builds
- [x] read diff carefully
- [x] any other backports we want/need?
- [x] perhaps suppress the testsuite errors from NumPy warnings deprecations with pre-release NumPy, `DeprecationWarning: NumPy warning suppression and assertion utilities are deprecated. Use warnings.catch_warnings, warnings.filterwarnings, pytest.warns, or pytest.filterwarnings instead. (Deprecated NumPy 2.4)
`; I don't think we're obligated to backport the actual fixes since these are just testsuite annoyances, and execution of the deprecation->removal should fall outside our support window

rgommers committed:
CI: bump OpenBLAS version to latest release (0.3.29.265.1)

MugundanMCW committed:
CI: validate build on Win-ARM64  (#23133)

* Add WOA support to windows.yml

Adding windows on ARM support for windows.yml file to build arm64 wheels
create action.yml to split LLVM and pkgconf changes as reusable blocks.

* TST: skip lapack and spatial.distance.cdist tests that crash on win-arm64

These are both caused by a crash in `np.linalg.env`, see numpy#29442.
The `scipy` functionality under test is fine, verified by separating out
the `linalg.inv` call from the scipy function calls in the problematic
CI job.

Co-authored-by: Ralf Gommers <ralf.gommers@gmail.com>

khmyznikov committed:
Merge pull request #23122 from khmyznikov/main

BLD: Windows Arm64 wheel support

nickodell committed:
BUG: signal: Avoid unbounded cache for ShortTimeFFT

If @functools.cache() is used on an instance method, this
will hold references to all instances that were provided
as the `self` argument.

Greenie0701 committed:
BUG: Skip installing gmpy2(test deps) for Win-ARM64 (#23403)

rgommers committed:
MAINT: ensure Qhull license file gets included in wheels (#23435)

rgommers committed:
BUG: fix regression for non-integer `maxiter` in SLSQP (#23442)

* MAINT: optimize: guard maxiter for noninteger values in SLSQP

* TST: regression test for non-integer maxiter for `slsqp` et al

* DOC: minor fix to `maxiter` docstring entry

---------

Co-authored-by: ilayn <ilhanpolat@gmail.com>

DietBru committed:
In `signal.shortTimeFFT`: remove typo in attribute name.

DimitriPapadopoulos committed:
MAINT: Update licensing of bundled software (#23452)

* Update licensing of bundled software
* Discard references to software removed from SciPy sources.
* Fix paths to bundled software and license files.
* Name licenses consistently, use SPDX short identifiers:
  - https://spdx.org/licenses/BSD-2-Clause.html
  - https://spdx.org/licenses/BSD-3-Clause.html
  - https://spdx.org/licenses/BSL-1.0.html

nabenabe0928 committed:
BUG:optimize: Fix the incompatible dtypes in L-BFGS-B wrapper (#23471)

adammj committed:
BUG: signal.lombscargle: replace in-place input adjustments (#23475)

See gh-23474

* Add test to check for mutation of any of the 4 input arrays, and source
code fixes that allow the new test to pass.

nickodell committed:
MAINT: Bump CuPy to 13.6.0

nickodell committed:
MAINT: Use 24.04 to get nvidia-driver-580

The Ubuntu 22.04 image has nvidia-driver-570, which is not compatible
with CUDA 13.0.

JostMigenda committed:
MAINT: add lower bound for pytest (#23495)

* MAINT: add lower bound for pytest

pytest-7.x failed to emit warnings, which is fixed in pytest-8.0.0
Closes gh-22244.

* MAINT: also add pytest bound to pyproject.toml/requirements.txt

tylerjereddy committed:
MAINT, BLD: ignore meson wraplock (#23540)

* The `1.9.0` release of `meson` from a few days
ago includes some locking improvements that consistently
lead to at least one `.wraplock` file artifact in the SciPy
directory when using `spin build ...` commands. I think
we might as well ignore these files just like they do
upstream at https://github.com/mesonbuild/meson/pull/13854.

[ci skip] [skip ci]

tylerjereddy committed:
BUG: fixups after backporting

DietBru commented:
> 8. https://github.com/scipy/scipy/pull/23457

It seems like from PR #23457 only  the commit  "[In signal.shortTimeFFT: remove typo in attribute name.](https://github.com/scipy/scipy/pull/23457/commits/5f65699b186faf92691996c444c6349676f34009)"  
is included and not the commit "[BUG: signal.ShortTimeFFT: Remove @lru_cache decorators](https://github.com/scipy/scipy/pull/23457/commits/6da75af8424426f4a6c7e7aa73dd31b8f089688b)". 
Was that intended?

tylerjereddy committed:
TST, MAINT: PR 23543 revisions

* Suppress NumPy deprecation warnings related to NumPy
warning suppression utilities, since supporting that
shouldn't be needed on the release branch.

tylerjereddy committed:
BUG: PR 23543 revisions

* Attempt to fix up failure to properly cherry-pick/backport
content from gh-23457.

* I effectively checked out the two files touched in that PR
directly from `main` at the time of writing.

tylerjereddy commented:
I pulled in another backport to bump the version of OpenBLAS, needed for Windows ARM support.

Let's see where the CI settles--pretty sure there are still a few mysteries to sort out.

DietBru commented:
> @DietBru
> 
> > Was that intended?
> 
> It looks like that PR was squash-merged--what patch do you see locally for the squashed commit with `git show cad5e3b`? That looks like quite a small patch to me locally that doesn't include what I need? I must be missing something here... or the order of merges/merge conflicts on related PRs messed something up.

Not only commit cad5e3b is needed but also the previous commit 51c848eb0f (I am also not exactly sure what happened).
 
> For now, I've tried to clean it up by checking out the related files directly from `main` -- can you take a look to see if the diff is more to your liking now? I believe I may have pulled in an extra static typing changeset we may not need/want perhaps.

It looks fine now—all `@cache` / `@lru_cache` decorators (which were leaking memory) are now removed in the `ShortTimeFFT` class. 

Just for the record: PR #23118 also slipped in, which just added the following line to class  `ShortTimeFFT`:
```python
class ShortTimeFFT:
   ...
    # generic type compatibility with scipy-stubs
    __class_getitem__ = classmethod(GenericAlias)
```
with a unit test in file `test_short_time_fft.py`. This will not have any visible effect on users as far as I can see.

DietBru commented:
The CI test failures [no phytran & sdist],  [mypy (py3.11)] in `test_signaltools.py` are caused by this 
[test]:
```python
@pytest.mark.filterwarnings('ignore::DeprecationWarning')
@skip_xp_backends(np_only=True, reason="object arrays")
class TestLinearFilterDecimal(_TestLinearFilter):
    dtype = np.dtype('O')

    def type(self, x):
        return Decimal(str(x))


@pytest.mark.filterwarnings('ignore::DeprecationWarning')
@skip_xp_backends(np_only=True, reason="object arrays")
class TestLinearFilterObject(_TestLinearFilter):
    dtype = np.dtype('O')
    type = float
```
They use `object` as dtype. It's support (and tests) was removed in PR #23212. Perhaps disabling those tests might be sufficient.

[no phytran & sdist]: https://github.com/scipy/scipy/actions/runs/17479428560/job/49646731592?pr=23543
[mypy (py3.11)]: https://github.com/scipy/scipy/actions/runs/17479428571/job/49646731518?pr=23543
[test]: https://github.com/tylerjereddy/scipy/blob/963a466c8f6261227304c05b8974cb2b63da515e/scipy/signal/tests/test_signaltools.py#L2356

rgommers committed:
CI: pin scikit-umfpack to 0.3.3 to avoid failures, 0.4.2 is broken. (#23576)

tylerjereddy commented:
@rgommers do you have an opinion given the deep dive in https://github.com/numpy/numpy/issues/29721#issuecomment-3272541768? Probably wasn't a great use of time, but we have a few viable routes forward, either with or without NumPy cooperation, and most of them are pretty simple (NumPy changes 1 line, or we shim to accept an extra error class, or we skip the tests and reference appropriate tickets).

tylerjereddy committed:
TST: PR 23543 revisions

* Broadened the allowed error classes raised by two tests:
`TestLinearFilterDecimal::test_bad_size_zi` and
`TestLinearFilterObject::test_bad_size_zi`

* This is to compensate for an upstream change:
https://github.com/numpy/numpy/issues/29721

rgommers committed:
BUG: fix memory leaks in `optimize.nnls` and `linalg.sqrtm` (#23563)

* BUG: optimize: fix memory leak in NNLS wrapper
* BUG: linalg: fix memory mgmt issue in sqrtm C code

---------

Co-authored-by: ilayn <ilhanpolat@gmail.com>

tylerjereddy committed:
DOC, MAINT: PR 23543 revisions

* Update the SciPy `1.16.2` release notes/`.mailmap`.

* Back out gh-23118, which was not intended to be backported.

tylerjereddy commented:
I ended up just allowing the extra error class in the test--the way these old tests are written made that far more pragmatic.

We're now at parity with all merged backport PRs, and @jorenham I've backed out the typing change that snuck in.

I need to read the diff carefully, but let's first see if regular CI passes after the latest updates and go from there.

tylerjereddy committed:
TST: PR 23543 revisions

* Suppress the problem test case described at:
https://github.com/scipy/scipy/pull/23543#issuecomment-3276286172

tylerjereddy committed:
CI: PR 23543 revisions [wheel build]

* Empty commit to test wheel build matrix
prior to SciPy `1.16.2` release process.

[wheel build]
